### PR TITLE
6579 Allow for FHIR to HL7 schema extensions

### PR DESCRIPTION
**Pull request by carlosfelix2**
This PR allows for FHIR to HL7 schema extensions.

Test Steps:
1. Build this branch.
2. Run the following command to generate an HL7 using the new CO extended schema: `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir -s metadata/hl7_mapping/STLTs/CO/CO.yml`
3. Compare the output to `prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.hl7` and verify the overridden elements and HL7 fields are as defined in the CO.yml file

## Changes
- Created initial CO schema with data from the HCA's CO requirements plus RS CO settings
- Schemas can be extended to overwrite elements by name or add more elements.

### Testing
- [ ] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #6579

### CODE REVIEW

File: prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ elements:
     value: [ '%resource.extension(%`rsext-order-control`).value.coding.code' ]
     hl7Spec: [ '%{hl7Order}-1' ]
 
-  - name: placer-order-number
+  - name: order-placer-order-number
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml:11]: Note: Name change is to make sure the elements are unique

File: prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml [hunk 2]
```diff
@@ -1,39 +1,49 @@
...
 
-  - name: organization-id-type
+  - name: xon-organization-id-type
     resource: '%resource.type.where(coding.system = ''http://terminology.hl7.org/CodeSystem/v2-0203'')'
     condition: '%resource.exists()'
     value: [ '%resource.coding[0].code' ]
     hl7Spec: [ '%{hl7OrgField}-7' ]
 
-  - name: organization-id
+  - name: xon-assigning-facility
```
Comment (carlosfelix2) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml:37]: Note: Added these, so I can override them in the CO one.

**Comment by PatriciaPerozoUSDS** [comment]
I'm getting errors when I run the test command you added. (I cleaned and packaged after the first time seeing errors)

I see: 
```
17:14:30.587 [main][1] TRACE gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter - Started processing of element co-ordering-facility-assigning-facility...
Exception in thread "main" org.hl7.fhir.r4.utils.FHIRLexer$FHIRLexerException: Error in ?? at 1, 1: Path cannot be empty
        at org.hl7.fhir.r4.utils.FHIRLexer.error(FHIRLexer.java:145)
        at org.hl7.fhir.r4.utils.FHIRLexer.error(FHIRLexer.java:141)
        at org.hl7.fhir.r4.utils.FHIRPathEngine.parse(FHIRPathEngine.java:337)
        at org.hl7.fhir.r4.utils.FHIRPathEngine.parse(FHIRPathEngine.java:331)
        at gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils.parsePath(FhirPathUtils.kt:51)
        at gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils.evaluateString(FhirPathUtils.kt:127)
        at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.getValue$prime_router(FhirToHl7Converter.kt:138)
        at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processElement$prime_router(FhirToHl7Converter.kt:108)
        at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema$prime_router(FhirToHl7Converter.kt:75)
        at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema$prime_router$default(FhirToHl7Converter.kt:65)
        at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.convert(FhirToHl7Converter.kt:58)
        at gov.cdc.prime.router.cli.ProcessFhirCommands.run(ProcessFhirCommands.kt:84)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
        at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
        at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
        at gov.cdc.prime.router.cli.MainKt.main(main.kt:280)
```
I'm wondering if its an order of processing issue? Maybe its trying to add to `/PATIENT_RESULT/ORDER_OBSERVATION/ORC` before it exists?

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+name: CO
+hl7Type: ORU_R01
+hl7Version: 2.5.1
+extends: ../../ORU_R01/ORU_R01-base
+elements:
+  # Source: ReportStream CO Requirements from settings
+  - name: receiving-application-namespace-id
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml:7]: What's the purpose of having this name and value if it doesn't have an hl7Spec or isn't referenced elsewhere in the ORU_R01 schema?

Comment (PatriciaPerozoUSDS) [reply prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml:7]: Wait I take it back! I researched and found it referenced in MSH my bad!

**Comment by PatriciaPerozoUSDS** [comment]
> I'm getting errors when I run the test command you added. (I cleaned and packaged after the first time seeing errors)
> 
> I see:
> 
> ```
> 17:14:30.587 [main][1] TRACE gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter - Started processing of element co-ordering-facility-assigning-facility...
> Exception in thread "main" org.hl7.fhir.r4.utils.FHIRLexer$FHIRLexerException: Error in ?? at 1, 1: Path cannot be empty
>         at org.hl7.fhir.r4.utils.FHIRLexer.error(FHIRLexer.java:145)
>         at org.hl7.fhir.r4.utils.FHIRLexer.error(FHIRLexer.java:141)
>         at org.hl7.fhir.r4.utils.FHIRPathEngine.parse(FHIRPathEngine.java:337)
>         at org.hl7.fhir.r4.utils.FHIRPathEngine.parse(FHIRPathEngine.java:331)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils.parsePath(FhirPathUtils.kt:51)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils.evaluateString(FhirPathUtils.kt:127)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.getValue$prime_router(FhirToHl7Converter.kt:138)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processElement$prime_router(FhirToHl7Converter.kt:108)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema$prime_router(FhirToHl7Converter.kt:75)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.processSchema$prime_router$default(FhirToHl7Converter.kt:65)
>         at gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter.convert(FhirToHl7Converter.kt:58)
>         at gov.cdc.prime.router.cli.ProcessFhirCommands.run(ProcessFhirCommands.kt:84)
>         at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
>         at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
>         at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
>         at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
>         at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
>         at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
>         at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
>         at gov.cdc.prime.router.cli.MainKt.main(main.kt:280)
> ```
> 
> I'm wondering if its an order of processing issue? Maybe its trying to add to `/PATIENT_RESULT/ORDER_OBSERVATION/ORC` before it exists?

Oh I figured out what the issue is - I'll add inline comments in the code about whats breaking on my machine

File: prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml [hunk 4]
```diff
@@ -38,25 +38,30 @@ elements:
...
-  - name: transaction-time
+  - name: order-transaction-time
     resource: '%resource.basedOn.resolve()'
     value: [ '%resource.authoredOn' ]
     hl7Spec: [ '%{hl7Order}-9' ]
 
-  - name: ordering-provider
+  - name: order-entered-by
+    condition: false
+    value: [ '' ] # Don't have data yet
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml:54]: This needs to be `''''''` instead of just `''`. Its being evaluated as a json path instead of an empty string and throwing an error.

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  # Source: HCA CO Requirements
+  # Always replace ORC-21.8 with NULL
+  - name: co-ordering-facility-assigning-facility
+    value: [ '' ]
+    hl7Spec: [ '/PATIENT_RESULT/ORDER_OBSERVATION/ORC-21-8' ]
+
+  # Source: HCA CO Requirements
+  # Always replace ORC-21.9 with NULL
+  - name: co-ordering-facility-name-code
+    value: [ '' ]
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml:44]: This needs to be '''''' instead of just ''. It's being evaluated as a json path instead of an empty string and throwing an error.

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  # Reporting facility name: CO requested to keep original MSH-4 from HCA.  ReportStream hard codes it to
+  # CDC PRIME^11D2030855^CLIA
+  # TODO - How does this change for other senders?
+
+  # TODO - These are HCA specific configs that should go in the sender transformation, but these may be CO specific too
+  # Source: HCA CO Requirements
+  # Always replace ORC-21.8 with NULL
+  - name: co-ordering-facility-assigning-facility
+    value: [ '' ]
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml:38]: This needs to be '''''' instead of just ''. It's being evaluated as a json path instead of an empty string and throwing an error.

File: prime-router/metadata/hl7_mapping/ORU_R01/base/message-header.yml [hunk 7]
```diff
@@ -1,13 +1,13 @@
 name: ORU-R01-message-header
 elements:
   - name: sending-application-namespace-id
-    resource: '%resource.source.extension(%`rsext-source-software-vendor-org`).value.resolve()'
-    value: [ '%resource.extension(%`rsext-assigning-authority-namespace-id`).value' ]
+    # Original value is in %resource.extension(%`rsext-assigning-authority-namespace-id`).value
+    value: [ '''CDC PRIME - Atlanta, Georgia (Dekalb)''' ]
```
Comment (clediggins-usds) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/message-header.yml:5]: Why are we hard coding?

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml (hunk 5 above)
Comment (clediggins-usds) [reply prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml:44]: Getting the same error

**Review by clediggins-usds — CHANGES_REQUESTED** [review]

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml (hunk 3 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml:7]: This one is overwriting an existing element in the parent schema

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml (hunk 6 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml]: fixed.  thx

File: prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml (hunk 5 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/STLTs/CO/CO.yml]: fixed.  thx

File: prime-router/metadata/hl7_mapping/ORU_R01/base/message-header.yml (hunk 7 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/message-header.yml]: This is what is done in ReportStream today.  It tells the receivers the data is coming from ReportStream instead of from an entity directly.

File: prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml (hunk 4 above)
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml:54]: Even though the condition is false, can you update these as well?

File: prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml [hunk 8]
```diff
@@ -1,35 +1,45 @@
...
+    condition: false # We have no data for this field yet
+    value: [ '' ]
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml:35]: Same here

File: prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml [hunk 9]
```diff
@@ -1,35 +1,45 @@
...
+    hl7Spec: [ '%{hl7OrgField}-8' ]
+
+  - name: xon-name-representation-code
+    condition: false # We have no data for this field yet
+    value: [ '' ]
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml:40]: Here as well

File: prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml (hunk 4 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/order.yml:54]: Fixed the code to allow for empty FHIR paths, so it does not crash

File: prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml (hunk 8 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml:35]: Fixed the code to allow for empty FHIR paths, so it does not crash

File: prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml (hunk 9 above)
Comment (carlosfelix2) [reply prime-router/metadata/hl7_mapping/ORU_R01/base/xon-organization.yml:40]: Fixed the code to allow for empty FHIR paths, so it does not crash

**Review by PatriciaPerozoUSDS — COMMENTED** [review]
Tested locally everything is working and makes sense - do you mind adding unit tests to document the empty path changes?

**Review by PatriciaPerozoUSDS — APPROVED** [review]

**Review by clediggins-usds — APPROVED** [review]
